packages/hash/api/src/db/postgres/aggregation/sql/aggregations.util.ts

```diff
@@ -0,0 +1,234 @@
+import { sql, ValueExpressionType } from "slonik";
+
+import { Connection } from "../../types";
+import { DbAggregation } from "../../../adapter";
+import { mapColumnNamesToSQL } from "../../util";
+import {
+  aggregationVersionsColumnNames,
+  DbAggregationVersionRow,
+  insertAggregationVersionRow,
+} from "./aggregation_versions.util";
+
+export const aggregationsColumnNames = [
+  "aggregation_id",
+  "path",
+  "source_account_id",
+  "source_entity_id",
+  "applied_to_source_at",
+  "applied_to_source_by_account_id",
+  "removed_from_source_at",
+  "removed_from_source_by_account_id",
+];
+
+export const aggregationsColumnNamesSQL = mapColumnNamesToSQL(
+  aggregationsColumnNames,
+);
+
+export type DbAggregationRow = {
+  aggregation_id: string;
+  source_account_id: string;
+  source_entity_id: string;
+  path: string;
+  applied_to_source_at: string;
+  applied_to_source_by_account_id: string;
+  removed_from_source_at: string | null;
+  removed_from_source_by_account_id: string | null;
+};
+
+type DbAggregationWithVersionRow = DbAggregationRow & DbAggregationVersionRow;
+
+/** selects all aggregation versions in the datastore */
+const selectAllAggregationVersions = sql<DbAggregationWithVersionRow>`
+  select
+    ${mapColumnNamesToSQL(aggregationsColumnNames, "aggregations")},
+    ${mapColumnNamesToSQL(
+      aggregationVersionsColumnNames.filter(
+        (column) => !aggregationsColumnNames.includes(column),
+      ),
+      "aggregation_versions",
+    )}
+  from
+    aggregations
+    join aggregation_versions
+    on aggregations.aggregation_id = aggregation_versions.aggregation_id
+`;
+
+/** selects all aggregations in the datastore with their latest version */
+const selectAllLatestVersionsOfAggregations = sql<DbAggregationWithVersionRow>`
+  select distinct on (aggregation_id) *
+  from (${selectAllAggregationVersions}) as all_aggregation_versions
+  order by aggregation_id, updated_at desc
+`;
+
+/** select all aggregations with the latest version before a given timestamp */
+const selectAllLatestVersionsOfAggregationsBeforeTimestamp = (params: {
+  beforeTimestamp: Date;
```

Comment:
"before" suggests less-than to me, rather than less-than-or-equal. I'm not sure what to do about that... "until" maybe?

Reply:
I agree, perhaps `beforeOrAtTimestamp ` would be more explicit?

Reply:
`beforeOrAtTimestamp` sounds good